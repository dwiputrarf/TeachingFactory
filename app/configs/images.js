/* eslint-disable global-require */
const images = {
  // back: require('../../assets/images/back.png'),
  // checkBoxOn: require('../../assets/images/checkbox_on.png'),
  // checkBoxOff: require('../../assets/images/checkbox_off.png'),
  // close: require('../../assets/images/close.png'),
  // dropdown: require('../../assets/images/ic_dropdown.png'),
  // faceSad: require('../../assets/images/face_sad.svg'),
  // login: require('../../assets/images/login.png'),
  // modalCamera: require('../../assets/images/modal_camera.png'),
  // modalGallery: require('../../assets/images/modal_gallery.png'),
  // noConnection: require('../../assets/images/no_connection.png'),
  // onBoarding: [
  //   require('../../assets/images/onBoarding_0.png'),
  //   require('../../assets/images/onBoarding_1.jpg'),
  //   require('../../assets/images/onBoarding_2.png')
  // ],
  // passOff: require('../../assets/images/padlock_off.png'),
  // passOn: require('../../assets/images/padlock_on.png'),
  // search: require('../../assets/images/ic_search.png'),
  // userOff: require('../../assets/images/user_off.png'),
  // userOn: require('../../assets/images/user_on.png')
};

export default images;
