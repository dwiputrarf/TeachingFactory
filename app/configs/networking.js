/* global fetch:false */

export const config = {
  baseUrl: 'https://reqres.in/api'
};

export const STATUS_CODE = {
  OK: 200,
  NOT_FOUND: 404,
  NO_CONTENT: 204
};

const fetchData = async (url, params, customHeaders) => {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders
  };

  // const tokens = await tokenService.get();
  // if (tokens) {
  //   headers = {
  //     ...headers,
  //     Authorization: `Bearer ${tokens.accessToken}`
  //   };
  // }

  const response = await fetch(url, {
    ...params,
    headers
  });
  // for DELETE method case
  if (response.status === STATUS_CODE.NO_CONTENT) return {};
  const json = await response.json();
  // if (!response.ok) {
  //   if (json && json.message) {
  //     if (json.data) {
  //       const error = new Error(json.message);
  //       error.code = json.data.errorCode;
  //       throw error;
  //     } else {
  //       throw new Error(json.message);
  //     }
  //   }
  //   throw new Error('unknownError');
  // }
  return json;
};

const get = async (endpoint, params = {}, headers = {}) => {
  let queryString = Object.keys(params)
    .map(key => `${key}=${params[key]}`)
    .join('&');
  if (queryString.length > 0) {
    queryString = `?${queryString}`;
  }
  const url = `${config.baseUrl}${endpoint}${queryString}`;
  const fetchParams = {
    method: 'GET'
  };
  return fetchData(url, fetchParams, headers);
};

const post = async (endpoint, params = {}, headers = {}) => {
  const url = `${config.baseUrl}${endpoint}`;
  const fetchParams = {
    method: 'POST',
    body: JSON.stringify(params)
  };
  return fetchData(url, fetchParams, headers);
};

const patch = async (endpoint, params = {}, headers = {}) => {
  const url = `${config.baseUrl}${endpoint}`;
  const fetchParams = {
    method: 'PATCH',
    body: JSON.stringify(params)
  };
  return fetchData(url, fetchParams, headers);
};

const put = async (endpoint, params = {}, headers = {}) => {
  const url = `${config.baseUrl}${endpoint}`;
  const fetchParams = {
    method: 'PUT',
    body: JSON.stringify(params)
  };
  return fetchData(url, fetchParams, headers);
};

const remove = async (endpoint, headers = {}) => {
  const url = `${config.baseUrl}${endpoint}`;
  const fetchParams = {
    method: 'DELETE'
  };
  return fetchData(url, fetchParams, headers);
};

export { get, post, put, patch, remove };
