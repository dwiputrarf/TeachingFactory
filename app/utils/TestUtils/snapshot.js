/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';

const mockSetState = Component =>
  class extends Component {
    setState(f, callback) {
      if (typeof f === 'function') {
        /* eslint-disable react/no-direct-mutation-state */
        this.state = { ...this.state, ...f(this.state) };
      } else {
        /* eslint-disable react/no-direct-mutation-state */
        this.state = { ...this.state, ...f };
      }
      if (callback) callback();
    }
  };

const assertSnapshot = (C, config) => {
  const { props, state, desc, throws, beforeTest } = config;

  let Component;
  let RenderedComponent;

  if (beforeTest) {
    beforeTest();
  }

  if (throws) {
    test(desc, () => {
      expect(() => {
        if (C.prototype.render) {
          Component = new C(props);
        } else {
          Component = C(props);
        }
      }).toThrow(throws);
    });
  } else {
    if (C.prototype.render) {
      const CWithMockState = mockSetState(C);
      Component = new CWithMockState(props);

      if (state) {
        Component.setState(state);
      }

      RenderedComponent = Component.render();
    } else {
      RenderedComponent = C(props);
    }

    test(desc, () => {
      const tree = RenderedComponent && renderer.create(RenderedComponent).toJSON();
      expect(tree).toMatchSnapshot();
    });
  }
};

const assertSnapshots = (Component, configs) => {
  configs.forEach(config => assertSnapshot(Component, config));
};

export { assertSnapshots, mockSetState };
export default { assertSnapshots, mockSetState };
