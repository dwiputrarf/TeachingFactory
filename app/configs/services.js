import { get, post } from './networking';

// end point
export const endpoint = {
  getListUser: async page => get(`/users?page=${page}`),
  getUserById: async id => get(`/users/${id}`),
  login: async params => post('/login', params)
};

export default { endpoint };
