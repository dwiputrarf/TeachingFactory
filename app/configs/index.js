import images from './images';
import { endpoint } from './services';

export const IMAGES = images;
export const ENDPOINT = endpoint;
