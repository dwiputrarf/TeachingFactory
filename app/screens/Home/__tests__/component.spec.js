import { assertSnapshots, mockSetState } from '../../../utils/TestUtils/snapshot';
import Component from '../component';
import { ENDPOINT } from '../../../configs';

jest.mock('../../../components/layouts/MainScreen', () => 'MainScreen');
jest.mock('../../../components/elements/Header', () => 'Header');
jest.mock('../../../components/elements/Button', () => 'Button');
jest.mock('../../../configs', () => ({
  ENDPOINT: {
    getListUser: jest.fn()
  }
}));
jest.mock('../../../utils/checkBandwidth', () => ({
  measureNetworkBandwitdh: jest.fn()
}));

describe('Component snapshot testing', () => {
  const configs = [
    {
      state: {},
      props: { listUsers: [] },
      desc: 'renders without data'
    },
    {
      state: { listUser: [{ id: 'id', first_name: '', last_name: 'last_name', avatar: 'avatar' }] },
      props: { listUsers: [] },
      desc: 'renders with data'
    }
  ];
  assertSnapshots(Component, configs);
});

describe('Component', () => {
  const ComponentWithMockedState = mockSetState(Component);
  const getInstance = props => new ComponentWithMockedState(props);
  beforeEach(() => {
    this.props = {
      actions: {
        fetchGetListUser: jest.fn()
      },
      listUsers: []
    };
  });

  afterEach(() => {
    this.props = null;
    jest.clearAllMocks();
  });

  test('componentDidMount', async () => {
    const instance = getInstance(this.props);
    await instance.componentDidMount();
    expect(this.props.actions.fetchGetListUser).toHaveBeenCalledWith('1');
  });

  describe('_loadData', async () => {
    test('success', async () => {
      const result = {
        data: [
          {
            id: 'id',
            first_name: 'first_name',
            last_name: 'last_name',
            avatar: 'avatar'
          }
        ]
      };
      ENDPOINT.getListUser.mockImplementationOnce(() => Promise.resolve(result));
      const instance = getInstance(this.props);
      await instance._loadData();
      expect(ENDPOINT.getListUser).toHaveBeenCalledWith('1');
    });

    test('failed', async () => {
      const error = { error: 'error' };
      ENDPOINT.getListUser.mockImplementationOnce(() => Promise.reject(error));
      const instance = getInstance(this.props);
      await instance._loadData();
      expect(ENDPOINT.getListUser).toHaveBeenCalledWith('1');
    });
  });
});
