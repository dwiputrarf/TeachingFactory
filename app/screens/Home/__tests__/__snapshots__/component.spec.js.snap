// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshot testing renders with data 1`] = `
<MainScreen
  isLoading={false}
>
  <Header
    back={true}
    setting={true}
    title="Home"
  />
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      }
    }
  />
</MainScreen>
`;

exports[`Component snapshot testing renders without data 1`] = `
<MainScreen
  isLoading={false}
>
  <Header
    back={true}
    setting={true}
    title="Home"
  />
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      }
    }
  />
</MainScreen>
`;
